exporters/elasticsearch-exporter/src/test/java/io/camunda/zeebe/exporter/BulkIndexRequestTest.java

```diff
@@ -0,0 +1,198 @@
+            + MAPPER.writeValueAsBytes(records.get(1)).length;
+    assertThat(request.memoryUsageBytes()).isEqualTo(expectedMemoryUsage);
+  }
+
+  @Test
+  void shouldClear() {
+    // given
+    final var records = recordFactory.generateRecords().limit(2).toList();
+    final var actions =
+        List.of(
+            new BulkIndexAction("index", "id", "routing"),
+            new BulkIndexAction("index2", "id2", "routing2"));
+    request.index(actions.get(0), records.get(0));
+    request.index(actions.get(1), records.get(1));
+
+    // when
+    request.clear();
+
+    // then
+    assertThat(request.bulkOperations()).isEmpty();
+    assertThat(request.isEmpty()).isTrue();
+    assertThat(request.memoryUsageBytes()).isEqualTo(0);
+    assertThat(request.size()).isEqualTo(0);
+    assertThat(request.lastIndexedMetadata()).isNull();
+  }
+
+  @Nested
+  final class IndexTest {
+    @Test
+    void shouldNotIndexWithIdenticalMetadata() {
+      // given
+      final var records = recordFactory.generateRecords().limit(2).toList();
+      final var action = new BulkIndexAction("index", "id", "routing");
+
+      // when - doesn't matter what the records are, if the metadata is the same we skip it
+      request.index(action, records.get(0));
+      request.index(action, records.get(1));
+
+      // then
+      assertThat(request.bulkOperations())
+          .extracting(BulkOperation::metadata)
+          .containsExactly(action);
+      assertThat(request.lastIndexedMetadata()).isEqualTo(action);
+      assertThat(request.isEmpty()).isFalse();
+    }
+
+    @Test
+    void shouldIndexWithDifferentMetadata() {
+      // given
+      final var records = recordFactory.generateRecords().limit(2).toList();
+      final var actions =
+          List.of(
+              new BulkIndexAction("index", "id", "routing"),
+              new BulkIndexAction("index2", "id2", "routing2"));
+
+      // when
+      request.index(actions.get(0), records.get(0));
+      request.index(actions.get(1), records.get(1));
+
+      // then
+      assertThat(request.bulkOperations())
+          .extracting(BulkOperation::metadata)
+          .containsExactlyElementsOf(actions);
+      assertThat(request.lastIndexedMetadata()).isEqualTo(actions.get(1));
+      assertThat(request.isEmpty()).isFalse();
+    }
+  }
+
+  @Nested
+  final class SerializationTest {
+    @Test
+    void shouldIndexRecordSerialized() {
+      // given
+      final var record = recordFactory.generateRecord();
+      final var action = new BulkIndexAction("index", "id", "routing");
+
+      // when
+      request.index(action, record);
+
+      // then
+      final var operations = request.bulkOperations();
+      assertThat(operations)
+          .hasSize(1)
+          .map(BulkOperation::metadata, this::deserializeSource)
+          .containsExactly(Tuple.tuple(action, record));
+    }
+
+    @Test
+    void shouldWriteOperationsAsNDJson() throws IOException {
```

Comment:
:thought_balloon: I thought about instead generating the expected serialized request and comparing both, but it is nicer on failure to see the deserialized version. However, of course, if deserialization completely failed then I guess it would be good to compare both in a text format, so open to discussion here.